File: src/storage.js
// storage.js - CSV and log file management

const fs = require('fs');
const path = require('path');

const DATA_DIR = path.join(__dirname, '..', 'data');
const LOGS_DIR = path.join(__dirname, '..', 'logs');

const DATA_FILE = path.join(DATA_DIR, 'silpo_raw.csv');
const LOG_FILE = path.join(LOGS_DIR, 'silpo_log.csv');

/**
 * Ensure directories and files exist
 */
function ensureLogFiles() {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }
  
  if (!fs.existsSync(LOGS_DIR)) {
    fs.mkdirSync(LOGS_DIR, { recursive: true });
  }
  
  // Create headers if files don't exist
  if (!fs.existsSync(DATA_FILE)) {
    const header = [
      'upload_ts', 'page_url', 'page_number', 'source', 'product_title', 'brand', 'product_type',
      'fat_pct', 'pack_qty', 'pack_unit', 'price_current', 'price_old', 'discount_pct',
      'price_per_l_or_kg_or_piece', 'rating', 'price_type'
    ];
    fs.writeFileSync(DATA_FILE, header.join(',') + '\n', 'utf-8');
  }
  
  if (!fs.existsSync(LOG_FILE)) {
    const header = ['ts', 'step', 'stage', 'message', 'url', 'status'];
    fs.writeFileSync(LOG_FILE, header.join(',') + '\n', 'utf-8');
  }
}

/**
 * Save products to CSV
 */
function saveToCSV(products) {
  if (!products || products.length === 0) {
    console.log('No products to save');
    return;
  }
  
  const rows = products.map(p => [
    p.upload_ts,
    p.page_url,
    p.page_number,
    p.source,
    escapeCSV(p.product_title),
    p.brand,
    p.product_type,
    p.fat_pct,
    p.pack_qty,
    p.pack_unit,
    p.price_current,
    p.price_old,
    p.discount_pct,
    p.price_per_l_or_kg_or_piece,
    p.rating,
    p.price_type,
  ]);
  
  // Append to file
  for (const row of rows) {
    fs.appendFileSync(DATA_FILE, row.join(',') + '\n', 'utf-8');
  }
  
  console.log(`Appended ${products.length} products to ${DATA_FILE}`);
}

/**
 * Append log entry
 */
function appendLog(logEntry) {
  const row = [
    logEntry.ts,
    logEntry.step,
    logEntry.stage,
    escapeCSV(logEntry.message),
    escapeCSV(logEntry.url),
    logEntry.status,
  ];
  
  fs.appendFileSync(LOG_FILE, row.join(',') + '\n', 'utf-8');
}

/**
 * Escape CSV values
 */
function escapeCSV(value) {
  if (!value) return '';
  const str = String(value);
  
  if (str.includes(',') || str.includes('"') || str.includes('\n')) {
    return '"' + str.replace(/"/g, '""') + '"';
  }
  
  return str;
}

/**
 * Get recent data
 */
function getRecentData(days = 1) {
  if (!fs.existsSync(DATA_FILE)) return [];
  
  const cutoffTime = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();
  const lines = fs.readFileSync(DATA_FILE, 'utf-8').split('\n');
  
  return lines
    .slice(1) // Skip header
    .filter(line => line.trim())
    .filter(line => {
      const ts = line.split(',')[0];
      return ts > cutoffTime;
    });
}

/**
 * Get statistics
 */
function getStatistics() {
  if (!fs.existsSync(DATA_FILE)) return null;
  
  const content = fs.readFileSync(DATA_FILE, 'utf-8');
  const lines = content.split('\n').filter(l => l.trim());
  
  return {
    totalRows: Math.max(0, lines.length - 1), // -1 for header
    lastUpdated: fs.statSync(DATA_FILE).mtime,
    fileSizeKB: fs.statSync(DATA_FILE).size / 1024,
  };
}

module.exports = {
  saveToCSV,
  appendLog,
  ensureLogFiles,
  getRecentData,
  getStatistics,
  DATA_FILE,
  LOG_FILE,
};

File: src/storage.py
"""CSV and log file management - Production Grade"""

import csv
import logging
from pathlib import Path
from datetime import datetime
from typing import List, Dict
from src import config

logger = logging.getLogger(__name__)


class Storage:
    """CSV file management for scraper data"""
    
    @staticmethod
    def ensure_directories():
        """Ensure data and logs directories exist"""
        config.DATA_DIR.mkdir(parents=True, exist_ok=True)
        config.LOGS_DIR.mkdir(parents=True, exist_ok=True)
        logger.info(f"✓ Directories ensured: {config.DATA_DIR}, {config.LOGS_DIR}")
        
        # Create headers if files don't exist
        if not config.DATA_FILE.exists():
            Storage.create_data_file()
        
        if not config.LOG_FILE.exists():
            Storage.create_log_file()
    
    @staticmethod
    def create_data_file():
        """Create data CSV with headers"""
        try:
            with open(config.DATA_FILE, 'w', newline='', encoding='utf-8') as f:
                writer = csv.DictWriter(f, fieldnames=config.CSV_HEADERS)
                writer.writeheader()
            logger.info(f"✓ Created {config.DATA_FILE}")
        except Exception as e:
            logger.error(f"Failed to create data file: {str(e)}")
    
    @staticmethod
    def create_log_file():
        """Create log CSV with headers"""
        headers = ['ts', 'step', 'stage', 'message', 'url', 'status']
        try:
            with open(config.LOG_FILE, 'w', newline='', encoding='utf-8') as f:
                writer = csv.DictWriter(f, fieldnames=headers)
                writer.writeheader()
            logger.info(f"✓ Created {config.LOG_FILE}")
        except Exception as e:
            logger.error(f"Failed to create log file: {str(e)}")
    
    @staticmethod
    def save_products(products: List[Dict]):
        """
        Append products to CSV
        
        Args:
            products: List of product dictionaries
        """
        if not products:
            logger.warning("No products to save")
            return
        
        try:
            with open(config.DATA_FILE, 'a', newline='', encoding='utf-8') as f:
                writer = csv.DictWriter(f, fieldnames=config.CSV_HEADERS)
                writer.writerows(products)
            logger.info(f"✓ Appended {len(products)} products to {config.DATA_FILE}")
        except Exception as e:
            logger.error(f"Failed to save products: {str(e)}")
    
    @staticmethod
    def append_log(log_entry: Dict):
        """
        Append log entry
        
        Args:
            log_entry: Dictionary with ts, step, stage, message, url, status
        """
        headers = ['ts', 'step', 'stage', 'message', 'url', 'status']
        
        try:
            with open(config.LOG_FILE, 'a', newline='', encoding='utf-8') as f:
                writer = csv.DictWriter(f, fieldnames=headers)
                writer.writerow(log_entry)
        except Exception as e:
            logger.error(f"Failed to append log: {str(e)}")
    
    @staticmethod
    def get_statistics() -> Dict:
        """Get file statistics"""
        if not config.DATA_FILE.exists():
            return None
        
        try:
            with open(config.DATA_FILE, 'r', encoding='utf-8') as f:
                lines = f.readlines()
            
            return {
                'total_rows': max(0, len(lines) - 1),  # -1 for header
                'last_updated': datetime.fromtimestamp(config.DATA_FILE.stat().st_mtime),
                'file_size_kb': round(config.DATA_FILE.stat().st_size / 1024, 2),
            }
        except Exception as e:
            logger.error(f"Failed to get statistics: {str(e)}")
            return None
